cli/flox-rust-sdk/src/providers/services.rs

```diff
@@ -405,4 +581,202 @@ mod tests {
+            write_process_compose_config(config, &config_path).unwrap();
+
+            let mut cmd = Command::new(&*PROCESS_COMPOSE_BIN);
+
+            // apparently it doesn't do this automatically even though it's not connected to a tty...
+            cmd.env("NO_COLOR", "1");
+            cmd.arg("--unix-socket")
+                .arg(&socket)
+                .arg("--config")
+                .arg(config_path)
+                .arg("--tui=false")
+                .arg("up")
+                .stdout(Stdio::null())
+                .stderr(Stdio::inherit());
+
+            // Dropping the child as stopping is handled via a process-compose command.
+            let child = cmd.spawn().unwrap();
+
+            let max_tries = 5;
+            for backoff in 1..max_tries {
+                println!("waiting for socket to exist");
+                thread::sleep(Duration::from_millis(100 * backoff));
+
+                // For now just check if the socket exists.
+                // Processes _may_ have not started yet, or the socket is unresponsive.
+                // We can't really check if the process is running,
+                // as it may have already exited.
+                // We could chek if the socket can be connected to
+                // or try to read ProcessStates, if the current approach leads to flaking tests.
+                if socket.exists() {
+                    break;
+                }
+
+                if backoff == max_tries {
+                    panic!("socket never appeared");
+                }
+            }
+
+            Self {
+                _temp_dir: temp_dir,
+                socket,
+                child,
+            }
+        }
+
+        /// Get the path to the socket.
+        fn socket(&self) -> &Path {
+            self.socket.as_ref()
+        }
+
+        /// Stop the `process-compose` instance.
+        fn stop(self) {
+            drop(self)
+        }
+    }
+
+    /// Try to stop the process-compose instance by sending a SIGTERM
+    /// to the process-compose process, which will stop all services.
+    impl std::ops::Drop for TestProcessComposeInstance {
+        fn drop(&mut self) {
+            let term_result = nix::sys::signal::kill(
+                nix::unistd::Pid::from_raw(self.child.id() as i32),
+                nix::sys::signal::SIGTERM,
+            );
+
+            if let Err(e) = term_result {
+                debug!("failed to send SIGTERM to process-compose: {:?}", e);
+            }
+        }
+    }
+
+    /// Test that [ProcessComposeLogReader] reads logs in order and sends them to the receiver.
+    #[test]
+    fn test_single_process_logs_received_in_order() {
+        let instance = TestProcessComposeInstance::start(&ProcessComposeConfig {
+            processes: [("foo".to_string(), ProcessConfig {
+                command: String::from(
+                    "i=0; while true; do i=$((i+1)); echo foo \"$((i))\"; sleep 0.1; done",
```

Comment:
nonblocking: do we need the sleep?

Reply:
without the sleep, at the time we start reading logs the logs are already in the 10-30kth iteration.

1) those logs dont seem to represent reality of most logging systems
2) you cant easily take the config and try it outside the test as the output would be overwhelming
3) the logs dont align with the asserted ones. we could change those to only check correct ordering but checking correct messages seems worthwhile imo.


That said, i think we could remove the sleep, but the behaviour is a bit nicer with some time to breathe.
If 0.1 sec are too "much" as it means the test takes a guaranteed Ω0.5 seconds, we can dial the sleep down a bit too.